### PR TITLE
Add diagnostics tab to sans gui

### LINKED ISSUE #21404: Add Diagnostics tab to sans GUI

### PR DESCRIPTION
**Pull request by Matthew-Andrew**
Description of work.
The diagnostic tab is a feature of the old SANS GUI which needs to be ported across to the new one.
In the old GUI this used the SumRowColumns Algorithm which only worked for LOQ. It has now been updated to use the ConvertToRealSpaceAxes and SumSpectra algorithms on order to cope with the tubes used by other SANS instruments.

**To test:**

- Code Review

- Find the loqdemo folder in the example data from here http://download.mantidproject.org/

- Open ISIS SANS V2 GUI
- Load the user files from loqdemo

- Switch to the diagnostic tab

- Select LOQ74044 as the file
- Click on Horizontal integral and verify that the below graph is plotted.

Fixes #21404 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [x] Is the code of an acceptable quality?
- [x] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [x] Are the unit tests small and test the class in isolation?
- [x] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [x] Do changes function as described? Add comments below that describe the tests performed?
- [x] Do the changes handle unexpected situations, e.g. bad input?
- [x] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SANS/sans/common/general_functions.py [hunk 1]
```diff
@@ -337,6 +337,97 @@ def is_part_of_reduced_output_workspace_group(state):
...
+    def _does_match(compiled_regex, line):
+        return compiled_regex.match(line) is not None
+
+    def _extract_simple_range(line):
+        start, stop = line.split("-")
+        start = int(start)
+        stop = int(stop)
+        if start > stop:
+            raise ValueError("Parsing event slices. It appears that the start value {0} is larger than the stop "
+                             "value {1}. Make sure that this is not the case.")
```
Comment (soininen) [inline scripts/SANS/sans/common/general_functions.py]: Formatting is missing the actual call to `.format()`.

File: scripts/SANS/sans/common/general_functions.py [hunk 2]
```diff
@@ -337,6 +337,97 @@ def is_part_of_reduced_output_workspace_group(state):
...
+
+    def _extract_simple_range_with_step_pattern(line):
+        start, stop = line.split("-")
+        start = int(start)
+        stop, step = stop.split(":")
+        stop = int(stop)
+        step = int(step)
+        if start > stop:
+            raise ValueError("Parsing event slices. It appears that the start value {0} is larger than the stop "
+                             "value {1}. Make sure that this is not the case.")
```
Comment (soininen) [inline scripts/SANS/sans/common/general_functions.py]: Missing `format()`?

File: scripts/SANS/sans/common/general_functions.py [hunk 3]
```diff
@@ -337,6 +337,97 @@ def is_part_of_reduced_output_workspace_group(state):
...
+        return range(start, stop+1, step)
+
+    def _extract_multiple_entry(line):
+        split_line = line.split(":")
+        start = int(split_line[0])
+        stop = int(split_line[1])
+        if start > stop:
+            raise ValueError("Parsing event slices. It appears that the start value {0} is larger than the stop "
+                             "value {1}. Make sure that this is not the case.")
```
Comment (soininen) [inline scripts/SANS/sans/common/general_functions.py]: Missing `format()`?

File: scripts/SANS/sans/gui_logic/models/diagnostics_page_model.py [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+                                            is_multi_range)
+        plot_graph(output_workspace)
+
+        output_workspaces.append(output_workspace)
+
+    return output_workspaces
+
+
+def integrate_ranges(ranges, integral, mask, detector, input_workspace_name, input_workspace, x_dim, y_dim, is_multi_range):
+    for range in ranges:
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/models/diagnostics_page_model.py]: `range` is a built-in function and being shadowed here. Actually, it seems to be shadowed in a lot of places below.

File: scripts/SANS/sans/gui_logic/presenter/diagnostic_presenter.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    def on_user_file_load(self, user_file):
+        self._view.user_file_name = user_file
+
+    def on_horizontal_clicked(self):
+        file = self._view.run_input
+        period = self._view.period
+        state_model_with_view_update = self._parent_presenter._get_state_model_with_view_update()
+        state = self._create_state(state_model_with_view_update, file, period, self._facility)
+        mask = self._view.horizontal_mask
+        range = self._view.horizontal_range
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/presenter/diagnostic_presenter.py]: `range` is a built-in.

File: scripts/SANS/sans/gui_logic/presenter/diagnostic_presenter.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        listener = DiagnosticsPagePresenter.IntegralListener(self)
+        detector = get_detector_from_gui_selection(self._view.detector)
+        self._work_handler.process(listener, self.run_integral, range, mask, IntegralEnum.Horizontal,
+                                   detector, state)
+
+    def on_vertical_clicked(self):
+        file = self._view.run_input
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/presenter/diagnostic_presenter.py]: `file` is a built-in function.

File: scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py [hunk 7]
```diff
@@ -233,7 +247,7 @@ def on_batch_file_load(self):
                 self._populate_row_in_table(row)
 
             # 5. Populate the selected instrument and the correct detector selection
-            self._setup_instrument_specific_settings()
+            #self._setup_instrument_specific_settings()
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py]: Commented-out code?

File: scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py [hunk 8]
```diff
@@ -245,14 +259,17 @@ def on_batch_file_load(self):
                                    " to the Mantid search directories! See here for more details: {}".format(str(e)))
 
     def on_data_changed(self):
-        # 1. Populate the selected instrument and the correct detector selection
-        self._setup_instrument_specific_settings()
+        # # 1. Populate the selected instrument and the correct detector selection
+        # self._setup_instrument_specific_settings()
```
Comment (soininen) [inline scripts/SANS/sans/gui_logic/presenter/run_tab_presenter.py]: Commented out code?

File: scripts/test/SANS/gui_logic/diagnostics_page_model_test.py [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+from sans.test_helper.mock_objects import create_mock_diagnostics_tab
+from sans.common.enums import SANSInstrument, DetectorType, IntegralEnum, SANSFacility
+from sans.gui_logic.models.diagnostics_page_model import create_state
+from sans.test_helper.mock_objects import (create_run_tab_presenter_mock)
+from sans.test_helper.user_file_test_helper import sample_user_file, create_user_file
+from sans.common.general_functions import (create_unmanaged_algorithm)
+from sans.gui_logic.models.state_gui_model import StateGuiModel
+from sans.user_file.user_file_reader import UserFileReader
+
+if sys.version_info.major == 3:
```
Comment (soininen) [inline scripts/test/SANS/gui_logic/diagnostics_page_model_test.py]: Would `sys.version_info.major > 2` be more forward compatible?

File: scripts/test/SANS/gui_logic/diagnostics_page_presenter_test.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
+from __future__ import (absolute_import, division, print_function)
+import unittest
+import sys
+from sans.test_helper.mock_objects import create_mock_diagnostics_tab
+from sans.common.enums import SANSInstrument, DetectorType, IntegralEnum, SANSFacility
+from sans.gui_logic.presenter.diagnostic_presenter import DiagnosticsPagePresenter
+from sans.test_helper.mock_objects import (create_run_tab_presenter_mock)
+
+if sys.version_info.major == 3:
```
Comment (soininen) [inline scripts/test/SANS/gui_logic/diagnostics_page_presenter_test.py]: Test `sys.version_info.major > 2`?

**Review by soininen — CHANGES_REQUESTED** [review]
Tested the new tab in the interface by following the provided instructions. Seemed to work fine, expect I couldn't compare the resulting plot to anything. Anyhow, the code changes look good to me, as well. I found just a few minor issues here and there.

**Comment by soininen** [comment]
@Mantid-Matthew The failing RHEL7 system test might be connected to these changes. Could you have a look?

**Comment by soininen** [comment]
@Mantid-Matthew The issue of commented-out line of code in `run_tab_presenter.py` still remains unresolved. Is there a reason to keep the line?

**Comment by Mantid-Matthew** [comment]
No I just managed to overlook the second of both the commented code and system comments. Thanks for spotting that.

**Comment by soininen** [comment]
@Mantid-Matthew Sorry for not noting this before, but how about release notes? Is there any other user facing documentation that needs to be updated?

**Comment by Mantid-Matthew** [comment]
No worries yes there should be some release notes and probably some sphinx documentation as well. Now you have reminded me this https://github.com/mantidproject/mantid/blob/master/docs/source/interfaces/ISIS%20SANS%20v2.rst needs updating to reflect the changes.

**Comment by Mantid-Matthew** [comment]
@soininen Documentation now added.

**Review by soininen — APPROVED** [review]

**Comment by martyngigg** [comment]
The failing Vesuvio tests are a known issue and they are unrelated to this issue.

**Pull request merged**